File: lib/leads/payment-handler.ts
/**
 * Lead Payment Handler
 * Processes payment.confirmed events from Chatwit/InfinitePay webhook
 * and manages manual payment registration.
 */

import { getPrismaInstance } from "@/lib/connections";
import { PaymentServiceType, PaymentStatus } from "@prisma/client";

const prisma = getPrismaInstance();

export interface PaymentConfirmedData {
	payment_link_id: number;
	order_nsu: string;
	amount_cents: number;
	paid_amount_cents: number;
	capture_method: string;
	receipt_url?: string;
	conversation_id: number;
	contact: {
		id: number;
		name: string;
		phone_number: string;
	};
}

export interface PaymentConfirmedPayload {
	event_type: "payment.confirmed";
	data: PaymentConfirmedData;
	metadata: {
		account_id: number;
		chatwit_base_url: string;
		timestamp: string;
	};
}

/**
 * Handles an incoming payment.confirmed event from Chatwit.
 * Looks up the lead by phone, creates a payment record, and auto-tags.
 */
export async function handlePaymentConfirmed(
	payload: PaymentConfirmedPayload,
	traceId: string,
): Promise<{ ok: boolean; leadId?: string; paymentId?: string; skipped?: boolean }> {
	const { data, metadata } = payload;

	// Idempotency: check if payment with this externalId already exists
	if (data.order_nsu) {
		const existing = await prisma.leadPayment.findUnique({
			where: { externalId: data.order_nsu },
		});
		if (existing) {
			console.log(`[PaymentHandler] Duplicate payment skipped: ${data.order_nsu}`, { traceId });
			return { ok: true, skipped: true, leadId: existing.leadId, paymentId: existing.id };
		}
	}

	// Look up lead by phone number (normalize to digits only for matching)
	const phoneDigits = data.contact.phone_number.replace(/\D/g, "");
	const lead = await prisma.lead.findFirst({
		where: {
			OR: [
				{ phone: { contains: phoneDigits } },
				{ sourceIdentifier: { contains: phoneDigits } },
			],
			source: "CHATWIT_OAB",
		},
		select: { id: true, tags: true },
	});

	if (!lead) {
		console.warn(`[PaymentHandler] No CHATWIT_OAB lead found for phone ${phoneDigits}`, { traceId });
		return { ok: true, skipped: true };
	}

	// Create payment record
	const payment = await prisma.leadPayment.create({
		data: {
			leadId: lead.id,
			amountCents: data.amount_cents,
			paidAmountCents: data.paid_amount_cents,
			serviceType: PaymentServiceType.OUTRO,
			status: PaymentStatus.CONFIRMED,
			captureMethod: data.capture_method,
			receiptUrl: data.receipt_url || null,
			externalId: data.order_nsu,
			confirmedAt: new Date(),
			confirmedBy: "webhook",
			chatwitConversationId: data.conversation_id,
			contactPhone: data.contact.phone_number,
			metadata: payload as any,
		},
	});

	// Auto-tag the lead
	const paymentTag = "pago";
	if (!lead.tags.includes(paymentTag)) {
		await prisma.lead.update({
			where: { id: lead.id },
			data: { tags: { push: paymentTag } },
		});
	}

	console.log(
		`[PaymentHandler] Payment recorded: ${payment.id} for lead ${lead.id} (R$ ${(data.paid_amount_cents / 100).toFixed(2)})`,
		{ traceId },
	);

	return { ok: true, leadId: lead.id, paymentId: payment.id };
}

File: platform-backend/domains/socialwise/services/flow/payment_handler.py
"""Payment confirmed handler for Socialwise payment webhook events."""

from __future__ import annotations

from typing import Any

from sqlalchemy import or_, select

from domains.socialwise.db.models.lead import Lead
from domains.socialwise.db.models.lead_payment import (
    LeadPayment,
    PaymentServiceType,
    PaymentStatus,
)
from domains.socialwise.db.session_compat import session_ctx
from domains.socialwise.services.flow.orchestrator import FlowOrchestrator
from platform_core.logging.config import get_logger

logger = get_logger(__name__)


async def handle_payment_confirmed(
    payload: dict[str, Any],
    trace_id: str | None = None,
) -> dict[str, Any]:
    data = payload.get("data") or {}
    contact = data.get("contact") or {}
    order_nsu = str(data.get("order_nsu") or "").strip()

    if order_nsu:
        async with session_ctx() as db:
            existing = (
                await db.execute(select(LeadPayment).where(LeadPayment.external_id == order_nsu))
            ).scalars().first()
            if existing:
                return {
                    "ok": True,
                    "skipped": True,
                    "leadId": existing.lead_id,
                    "paymentId": existing.id,
                }

    phone_digits = "".join(ch for ch in str(contact.get("phone_number") or "") if ch.isdigit())
    if not phone_digits:
        return {"ok": True, "skipped": True, "reason": "missing_contact_phone"}

    async with session_ctx() as db:
        lead = (
            await db.execute(
                select(Lead).where(
                    Lead.source == "CHATWIT_OAB",
                    or_(
                        Lead.phone.ilike(f"%{phone_digits}%"),
                        Lead.source_identifier.ilike(f"%{phone_digits}%"),
                    ),
                )
            )
        ).scalars().first()
        if not lead:
            logger.warning("payment_confirmed_lead_not_found", phone_digits=phone_digits, trace_id=trace_id)
            return {"ok": True, "skipped": True}

        payment = LeadPayment(
            lead_id=lead.id,
            amount_cents=int(data.get("amount_cents") or 0),
            paid_amount_cents=int(data.get("paid_amount_cents") or 0) or None,
            service_type=PaymentServiceType.OUTRO.value,
            status=PaymentStatus.CONFIRMED.value,
            capture_method=str(data.get("capture_method") or ""),
            receipt_url=data.get("receipt_url"),
            external_id=order_nsu or None,
            confirmed_by="webhook",
            chatwit_conversation_id=int(data.get("conversation_id") or 0) or None,
            contact_phone=str(contact.get("phone_number") or ""),
            metadata_json=payload,
        )
        db.add(payment)

        tags = list(lead.tags or [])
        if "pago" not in tags:
            lead.tags = [*tags, "pago"]

        await db.commit()
        await db.refresh(payment)

    try:
        conversation_id = str(data.get("conversation_id") or "").strip()
        if conversation_id and conversation_id != "0":
            orchestrator = FlowOrchestrator()
            await orchestrator.resume_from_payment(conversation_id, order_nsu, trace_id)
    except Exception as exc:
        logger.warning("payment_confirmed_resume_failed", error=str(exc), trace_id=trace_id)

    return {"ok": True, "leadId": lead.id, "paymentId": payment.id}
